src/NServiceBus.Core/Routing/AutomaticSubscriptions/AutoSubscribe.cs

```diff
@@ -59,26 +59,33 @@ static List<Type> GetMessageTypesHandledByThisEndpoint(MessageHandlerRegistry ha
 
         class ApplySubscriptions : FeatureStartupTask
         {
-            public ApplySubscriptions(IEnumerable<Type> messagesHandledByThisEndpoint)
+            public ApplySubscriptions(List<Type> messagesHandledByThisEndpoint)
             {
                 this.messagesHandledByThisEndpoint = messagesHandledByThisEndpoint;
             }
 
-            protected override async Task OnStart(IMessageSession session)
+            protected override Task OnStart(IMessageSession session)
             {
-                foreach (var eventType in messagesHandledByThisEndpoint)
+                var tasks = new Task[messagesHandledByThisEndpoint.Count];
+                for (var i = 0; i < messagesHandledByThisEndpoint.Count; i++)
                 {
-                    await session.Subscribe(eventType).ConfigureAwait(false);
-                    Logger.DebugFormat("Auto subscribed to event {0}", eventType);
+                    tasks[i] = SubscribeToEvent(session, messagesHandledByThisEndpoint[i]);
                 }
+                return Task.WhenAll(tasks);
             }
 
             protected override Task OnStop(IMessageSession session)
             {
                 return TaskEx.CompletedTask;
             }
 
-            IEnumerable<Type> messagesHandledByThisEndpoint;
+            static async Task SubscribeToEvent(IMessageSession session, Type eventType)
+            {
+                await session.Subscribe(eventType).ConfigureAwait(false);
+                Logger.DebugFormat("Auto subscribed to event {0}", eventType);
+            }
+
+            List<Type> messagesHandledByThisEndpoint;
```

Comment:
There's a bit less overhead in an array compared to a list, but this also isn't a per-message execution path, so I don't think it's worth changing.